### PR TITLE
issue 555.1: try loading previous data store snapshots if checkpoint file unavailable, enhance error message, address some warnings

### PR DESCRIPTION
**Pull request by vasiliy-arangodb**

### CODE REVIEW

File: arangod/Aql/EnumerateCollectionExecutor.cpp [hunk 1]
```diff
@@ -174,14 +174,14 @@ std::tuple<ExecutionState, EnumerateCollectionStats, size_t> EnumerateCollection
 
     if (_state == ExecutionState::WAITING) {
       return std::make_tuple<ExecutionState, EnumerateCollectionStats, size_t>( // tupple, cannot use initializer list due to build failure
-        ExecutionState(_state), std::move(stats), 0 // rvalue for first argument and std::move(...) required tmatch constructor signature
```
Comment (graetzer) [inline arangod/Aql/EnumerateCollectionExecutor.cpp]: isn't `ExecutionState(_state)` already an rvalue ?

File: arangod/Aql/EnumerateCollectionExecutor.cpp [hunk 2]
```diff
@@ -173,13 +173,18 @@ std::tuple<ExecutionState, EnumerateCollectionStats, size_t> EnumerateCollection
     std::tie(_state, _input) = _fetcher.fetchRow();
 
     if (_state == ExecutionState::WAITING) {
-      return {_state, stats, 0};
+      return std::make_tuple<ExecutionState, EnumerateCollectionStats, size_t>( // tupple, cannot use initializer list due to build failure
+        std::move(ExecutionState(_state)), std::move(stats), 0 // rvalue for first argument and std::move(...) required tmatch constructor signature
```
Comment (graetzer) [inline arangod/Aql/EnumerateCollectionExecutor.cpp]:
```suggestion
        ExecutionState(_state), std::move(stats), 0 // rvalue for first argument and std::move(...) required tmatch constructor signature
```

File: arangod/Aql/EnumerateCollectionExecutor.cpp [hunk 3]
```diff
@@ -173,13 +173,18 @@ std::tuple<ExecutionState, EnumerateCollectionStats, size_t> EnumerateCollection
...
+      );
     }
 
     if (!_input) {
       TRI_ASSERT(_state == ExecutionState::DONE);
-      return {_state, stats, 0};
+      return std::make_tuple<ExecutionState, EnumerateCollectionStats, size_t>( // tupple, cannot use initializer list due to build failure
+        std::move(ExecutionState(_state)), std::move(stats), 0 // rvalue for first argument and std::move(...) required tmatch constructor signature
```
Comment (graetzer) [inline arangod/Aql/EnumerateCollectionExecutor.cpp]:
```suggestion
        ExecutionState(_state), std::move(stats), 0 // rvalue for first argument and std::move(...) required tmatch constructor signature
```

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 4]
```diff
@@ -229,7 +229,9 @@ struct ExecuteSkipVariant<SkipVariants::FETCHER> {
   static std::tuple<ExecutionState, typename Executor::Stats, size_t> executeSkip(
       Executor& executor, typename Executor::Fetcher& fetcher, size_t toSkip) {
     auto res = fetcher.skipRows(toSkip);
-    return {res.first, typename Executor::Stats{}, res.second};
+    return std::make_tuple<ExecutionState, typename Executor::Stats, size_t>( // tupple, cannot use initializer list due to build failure
+      std::move(ExecutionState(res.first)), typename Executor::Stats{}, size_t(res.second) // rvalue required for all elements to match constructor signature
```
Comment (graetzer) [inline arangod/Aql/ExecutionBlockImpl.cpp]:
```suggestion
      ExecutionState(res.first), typename Executor::Stats{}, size_t(res.second) // rvalue required for all elements to match constructor signature
```

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 5]
```diff
@@ -250,7 +252,9 @@ struct ExecuteSkipVariant<SkipVariants::DEFAULT> {
     // this function should never be executed
     TRI_ASSERT(false);
     // Make MSVC happy:
-    return {ExecutionState::DONE, {}, 0};
+    return std::make_tuple<ExecutionState, typename Executor::Stats, size_t>( // tupple, cannot use initializer list due to build failure
+      std::move(ExecutionState(ExecutionState::DONE)), {}, 0 // rvalue required for all elements to match constructor signature
```
Comment (graetzer) [inline arangod/Aql/ExecutionBlockImpl.cpp]:
```suggestion
      ExecutionState(ExecutionState::DONE), {}, 0 // rvalue required for all elements to match constructor signature
```

File: arangod/Aql/IResearchViewExecutor.cpp [hunk 6]
```diff
@@ -208,17 +208,23 @@ IResearchViewExecutor<ordered>::skipRows(size_t toSkip) {
   if (!_inputRow.isInitialized()) {
     if (_upstreamState == ExecutionState::DONE) {
       // There will be no more rows, stop fetching.
-      return {ExecutionState::DONE, stats, 0};
+      return std::make_tuple<ExecutionState, typename IResearchViewExecutor<ordered>::Stats, size_t>( // tupple, cannot use initializer list due to build failure
+        std::move(ExecutionState(ExecutionState::DONE)), std::move(stats), 0 // rvalue required for all elements to match constructor signature
```
Comment (graetzer) [inline arangod/Aql/IResearchViewExecutor.cpp]:
```suggestion
        ExecutionState(ExecutionState::DONE), std::move(stats), 0 // rvalue required for all elements to match constructor signature
```

File: arangod/Aql/IResearchViewExecutor.cpp [hunk 7]
```diff
@@ -208,17 +208,23 @@ IResearchViewExecutor<ordered>::skipRows(size_t toSkip) {
...
+      );
     }
 
     std::tie(_upstreamState, _inputRow) = _fetcher.fetchRow();
 
     if (_upstreamState == ExecutionState::WAITING) {
-      return {_upstreamState, stats, 0};
+      return std::make_tuple<ExecutionState, typename IResearchViewExecutor<ordered>::Stats, size_t>( // tupple, cannot use initializer list due to build failure
+        std::move(ExecutionState(_upstreamState)), std::move(stats), 0 // rvalue required for all elements to match constructor signature
```
Comment (graetzer) [inline arangod/Aql/IResearchViewExecutor.cpp]:
```suggestion
        ExecutionState(_upstreamState), std::move(stats), 0 // rvalue required for all elements to match constructor signature
```

File: arangod/Aql/EnumerateCollectionExecutor.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/Aql/EnumerateCollectionExecutor.cpp]: these are rvalues already

**Comment by vasiliy-arangodb** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/4396/

File: arangod/Aql/EnumerateCollectionExecutor.cpp (hunk 1 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/EnumerateCollectionExecutor.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

File: arangod/Aql/EnumerateCollectionExecutor.cpp (hunk 2 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/EnumerateCollectionExecutor.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

File: arangod/Aql/EnumerateCollectionExecutor.cpp (hunk 3 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/EnumerateCollectionExecutor.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

File: arangod/Aql/ExecutionBlockImpl.cpp (hunk 4 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/ExecutionBlockImpl.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

File: arangod/Aql/ExecutionBlockImpl.cpp (hunk 5 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/ExecutionBlockImpl.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

File: arangod/Aql/IResearchViewExecutor.cpp (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/IResearchViewExecutor.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

File: arangod/Aql/IResearchViewExecutor.cpp (hunk 7 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/IResearchViewExecutor.cpp]: I would think so too, but MSVC was complaining with:
cannot convert argument 1 from 'arangodb::aql::ExecutionState' to 'arangodb::aql::ExecutionState &&'

Comment (mpoeter) [reply arangod/Aql/IResearchViewExecutor.cpp]: Try without explicitly specifying the template argument types:
```cpp
return std::make_tuple(ExecutionState(_upstreamState), std::move(stats), 0);
```
`make_tuple` uses universal references and perfect forwarding, but this requires that the argument type is deduced by the compiler (i.e., not explicitly specified).

Comment (vasiliy-arangodb) [reply arangod/Aql/IResearchViewExecutor.cpp]: @mpoeter removing the template arguments from std::make_tuple compiled everywhere, thank you

**Comment by vasiliy-arangodb** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/4412/

**Comment by vasiliy-arangodb** [comment]
replaced by https://github.com/arangodb/arangodb/pull/8975

**Pull request closed without merging**